tensorboard/uploader/proto/write_service.proto

```diff
@@ -190,6 +196,152 @@ message WriteTensorResponse {
+//    This response may imply that any provided data was ignored as redundant.
+//  * The client proceeds to upload from the specified offset (optionally
+//    validating the crc32c up to that point first).  Of course, if the object
+//    is already finalized, then nothing remains to upload, so the stream
+//    closes.
+//  * If the first request does provide data, that chunk may turn out to be
+//    redundant with a previous partial upload.  It is up to the client to
+//    weigh the tradeoff between this possibly-wasted transfer and an
+//    additional initial roundtrip to obtain the status.
+//
+//  * In general, the caller should not wait to verify that each response
+//    reports the expected status, size, and crc32c before sending the next
+//    request.  Doing so reduces throughput, with little benefit since any
+//    mismatch on the server side will cause the RPC to throw an error anyway.
+//  * Nonetheless the caller may wish to validate these values, if provided,
+//    whenever the response does arrive.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC. Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single request suffices (assuming no empty status
+//    request), so it is both the first and the last chunk.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+// Obtain a unique ID for a blob sequence, given the composite key
+// (experiment_id, run, tag, step).  If such a blob sequence already exists,
+// return its ID.  If not, create it first, and return the new ID.
+message GetOrCreateBlobSequenceRequest {
+  // Service-wide unique identifier of an uploaded log dir.
+  // eg: "1r9d0kQkh2laODSZcQXWP"
+  string experiment_id = 1;
+  // The name of the run to which the blob sequence belongs, for example
+  // "/some/path/mnist_experiments/run1/".
+  string run = 2;
+  // The name of the tag to which the blob sequence belongs, for example "loss".
+  string tag = 3;
+  // Step index within the run.
+  int64 step = 4;
+  // The total number of elements expected in the sequence.
+  // This effectively delares a number of initially empty 'upload slots',
+  // to be filled with subsequent WriteBlob RPCs.
+  int64 final_sequence_length = 5;
+  // Note that metadata.plugin_data.content does not carry the payload.
+  .tensorboard.SummaryMetadata metadata = 6;
+}
+
+message GetOrCreateBlobSequenceResponse {
+  // A unique ID for the the requested blob sequence.
+  string sequence_id = 1;
+}
+
+// A single chunk of the blob write stream.
+message WriteBlobRequest {
```

Comment:
We could add a field here for the total file size.  That way we could avoid sending a partial file when the whole thing would exceed quota anyway.

It'd be optional, and typically would produce a quota failure (if any) on the first request-- so long as the user is uploading only one blob at a time.  To handle this in the case of parallel write streams, we'd have to pre-reserve quota for the incoming blobs.  Maybe not worth the complexity, and the server-side implementation is postponable in any case.

For now, any opinions on whether a file size field is worth adding right away?

Reply:
I'd defer for now since it adds complexity and hitting the quota limit is intended to be a rare edge case.  If we change our minds, a strength of protobuf is that it's very easy to add fields later (and we basically just have to do a PR and a sync and then we can use it internally).